api/api.go

```diff
@@ -35,120 +30,115 @@ func HttpPOST(url string, data string) (resp *http.Response, err error) {
-// the API.
-func (srv *Server) handleHTTPRequest(mux *http.ServeMux, url string, handler http.HandlerFunc) {
-	mux.HandleFunc(url, func(w http.ResponseWriter, req *http.Request) {
-		if !strings.Contains(req.UserAgent(), srv.requiredUserAgent) {
+// requireUserAgent is middleware that requires all requests to set a
+// UserAgent that contains the specified string.
+func requireUserAgent(h http.Handler, ua string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if !strings.Contains(req.UserAgent(), ua) {
 			writeError(w, "Browser access disabled due to security vulnerability. Use Sia-UI or siac.", http.StatusBadRequest)
 			return
 		}
-		handler(w, req)
+		h.ServeHTTP(w, req)
 	})
 }
 
 // initAPI determines which functions handle each API call.
 func (srv *Server) initAPI() {
-	mux := http.NewServeMux()
-
-	// 404 Calls
-	srv.handleHTTPRequest(mux, "/", srv.unrecognizedCallHandler)
+	mux := httprouter.New()
+	mux.NotFound = http.HandlerFunc(srv.unrecognizedCallHandler) // custom 404
 
 	// Daemon API Calls - Unfinished
-	srv.handleHTTPRequest(mux, "/daemon/constants", srv.daemonConstantsHandler)
-	srv.handleHTTPRequest(mux, "/daemon/version", srv.daemonVersionHandler)
-	srv.handleHTTPRequest(mux, "/daemon/stop", srv.daemonStopHandler)
-	srv.handleHTTPRequest(mux, "/daemon/updates/apply", srv.daemonUpdatesApplyHandler)
-	srv.handleHTTPRequest(mux, "/daemon/updates/check", srv.daemonUpdatesCheckHandler)
+	mux.GET("/daemon/constants", srv.daemonConstantsHandler)
+	mux.GET("/daemon/version", srv.daemonVersionHandler)
+	mux.POST("/daemon/stop", srv.daemonStopHandler)
 
 	// Consensus API Calls
 	if srv.cs != nil {
-		srv.handleHTTPRequest(mux, "/consensus", srv.consensusHandler)            // GET
-		srv.handleHTTPRequest(mux, "/consensus/block", srv.consensusBlockHandler) // GET
+		mux.GET("/consensus", srv.consensusHandler)
+		mux.GET("/consensus/blocks/:height", srv.consensusBlocksHandler)
 	}
 
 	// Explorer API Calls
 	if srv.explorer != nil {
-		srv.handleHTTPRequest(mux, "/explorer", srv.explorerHandler)            // GET
-		srv.handleHTTPRequest(mux, "/explorer/", srv.explorerHandler)           // $(hash) GET
-		srv.handleHTTPRequest(mux, "/explorer/block", srv.explorerBlockHandler) // GET
+		mux.GET("/explorer", srv.explorerHandler)
+		mux.GET("/explorer/blocks/:height", srv.explorerBlocksHandler)
+		mux.GET("/explorer/hash/:hash", srv.explorerHashHandler)
 	}
 
 	// Gateway API Calls - Unfinished
 	if srv.gateway != nil {
-		srv.handleHTTPRequest(mux, "/gateway/status", srv.gatewayStatusHandler)
-		srv.handleHTTPRequest(mux, "/gateway/peers/add", srv.gatewayPeersAddHandler)
-		srv.handleHTTPRequest(mux, "/gateway/peers/remove", srv.gatewayPeersRemoveHandler)
+		mux.GET("/gateway", srv.gatewayHandler)
+		mux.POST("/gateway/add/:addr", srv.gatewayAddHandler)
+		mux.POST("/gateway/remove/:addr", srv.gatewayRemoveHandler)
```

Comment:
It's a more RESTful organization. One of the basic tenets is that resources should be uniquely identified by a URL, and should be manipulated by using different verbs on that URL. So it makes sense if you think of a peer as a resource.
It's not perfect though, because we have decided to only use GET and POST. The most RESTful organization would be:
```
PUT    /gateway/:addr (adds a peer)
GET    /gateway/:addr (accesses a peer)
DELETE /gateway/:addr (removes a peer)
```